src/full/Agda/TypeChecking/Substitute.hs

```diff
@@ -1633,6 +1638,16 @@ funSort' a b = case (a, b) of
   (a             , SizeUniv     ) -> sizeOfSort a >>= \case
     SmallSort{} -> Right SizeUniv
     LargeSort{} -> Left neverUnblock
+  -- No need to handle @LevelUniv@ in a special way here when --level-universe isn't on,
+  -- since this function is currently always called after reduction.
+  -- It would be safer to take it into account here, but would imply passing the option along as an argument.
+  (LevelUniv     , LevelUniv    ) -> Right LevelUniv
+  (a             , LevelUniv    ) -> sizeOfSort a >>= \case
+    SmallSort{} -> Left neverUnblock
+    LargeSort{} -> Right LevelUniv
```

Comment:
I don't think it should matter what the size of the sort is here, we should not allow any functions into `Level` unless the domain is in `LevelUniv`.

Reply:
Added note: some primitives in `Cubical.agda` such as `primTransp` make use of levels dependents on `I` terms, refusing `(IUniv, LevelUniv, LevelUniv)` as a PTS rule would make `--cubical` incompatible with `--level-universe`. Is this desirable ?

Reply:
The `I → Level` in transp and comp isn't essential but it does greatly simplify(!) the implementation of `transp` for data types and records. Since everything in the signature of a fibrant type must be fibrant, the implementation of `transp` for

```agda
data Foo Δ : Γ → Type ... where
```

looks like

```
transpFoo : (xs ys : Δ^I) → Foo (xs i0) (ys i0) → Foo (xs i1) (ys i1)
```

where we really do [exponentiate the whole telescope](https://github.com/agda/agda/blob/master/src/full/Agda/TypeChecking/Rules/Data.hs#L875). Since IUniv is already fairly restricted, essentially only having the interval (_and_ I plan to lock it down further!), I don't think the rule `(IUniv, LevelUniv, LevelUniv)` is problematic.

Reply:
Surely if levels can't depend on elements of types, they should also not be able to depend on interval variables?

Reply:
I would argue that the interval is (morally) special enough that it doesn't count as a type. But I see your point: we still run into a subject reduction issue as in

```agda
bad : SSetω
bad = (λ (ℓ : I → Level) → (i : I) → Set (ℓ i)) (λ _ → lzero)
```

One solution would be to add a rule `piSort I _ s2 = funSort IUniv (s2 i0)`; the arbitrary choice of `i0` requires postulates to observe, but it wouldn't be hard to patch the conversion checker to ignore interval terms in levels (add a flag to the `TCEnv`, set it in `equalLevel`, ignore `compareInterval` if the flag is set).

Reply:
One possible way to fix SR without restricting everything beyond rectognition would be to change the way whether something lives in a large sort is decided. Currently, if the codomain of a `piSort` has a free variable occuring in it, then the `piSort` lives in a large sort. However, it should be possible to get a more fine-grained typing rule now that `LevelUniverse` exists, meaning that the fact that `(l : Level) → Set l : Setω` would be inferred from the PTS Rule `(LevelUniv, Set, Setω)`. In such a setup, I believe `bad` here would live in `Set (lsuc lzero)`, as one would expect.